packages/@lwc/engine/src/framework/vm.ts

```diff
@@ -58,59 +68,81 @@ export enum VMState {
     disconnected,
 }
 
-export interface UninitializedVM {
-    /** Custom element tag name */
-    readonly tagName: string;
-    /** Component Element Back-pointer */
+export interface Context {
+    /** The attribute name used on the host element to scope the style. */
+    hostAttribute: string | undefined;
+    /** The attribute name used on all the elements rendered in the shadow tree to scope the style. */
+    shadowAttribute: string | undefined;
+    /** The VNode injected in all the shadow trees to apply the associated component stylesheets. */
+    styleVNode: VNode | null;
+    /** Object used by the template function to store information that can be reused between
+     *  different render cycle of the same template. */
+    tplCache: TemplateCache;
+    /** List of wire hooks that are invoked when the component gets connected. */
+    wiredConnecting: Array<() => void>;
+    /** List of wire hooks that are invoked when the component gets disconnected. */
+    wiredDisconnecting: Array<() => void>;
+}
+
+export interface VM {
+    /** The host element */
     readonly elm: HTMLElement;
-    /** Component Definition */
+    /** The host element tag name */
+    readonly tagName: string;
+    /** The component definition */
     readonly def: ComponentDef;
-    /** Component Context Object */
+    /** The component context object. */
     readonly context: Context;
-    /** Back-pointer to the owner VM or null for root elements */
+    /** The owner VM or null for root elements. */
     readonly owner: VM | null;
-    /** Component Creation Index */
-    idx: number;
-    /** Component state, analogous to Element.isConnected */
+    /** The component creation index. */
+    readonly idx: number;
+    /** The component connection state. */
     state: VMState;
-    data: VNodeData;
-    /** Shadow Children List */
+    /** The list of VNodes associated with the shadow tree. */
     children: VNodes;
-    /** Adopted Children List */
+    /** The list of adopted children VNodes. */
     aChildren: VNodes;
+    /** The list of custom elements VNodes currently rendered in the shadow tree. We keep track of
+     * those elements to efficiently unmount them when the parent component is disconnected without
+     * having to traverse the VNode tree. */
     velements: VCustomElement[];
-    cmpProps: Record<string, any>;
+    /** The component public properties. */
+    cmpProps: { [name: string]: any };
+    /** The mapping between the slot names and the slotted VNodes. */
     cmpSlots: SlotSet;
-    cmpFields: Record<string, any>;
-    callHook: (
-        cmp: ComponentInterface | undefined,
-        fn: (...args: any[]) => any,
-        args?: any[]
-    ) => any;
-    setHook: (cmp: ComponentInterface, prop: PropertyKey, newValue: any) => void;
-    getHook: (cmp: ComponentInterface, prop: PropertyKey) => any;
+    /** The component internal reactive properties. */
+    cmpFields: { [name: string]: any };
```

Comment:
Make it a record

Reply:
Discussed offline and @caridy is fine with this.